code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/DeploymentManagerImpl.java

```diff
@@ -268,12 +269,26 @@ public void doWithLockNoResult() {
 
     @Override
     public void activateGlobalServicesForHost(long accountId, long hostId) {
+
+        class DeferredServiceActivate implements Runnable {
+            Service service;
+
+            public DeferredServiceActivate(Service service) {
+                this.service = service;
+            }
+
+            @Override
+            public void run() {
+                activate(service);
+            }
+        }
+
         List<? extends Service> services = expMapDao.getActiveServices(accountId);
         for (Service service: services) {
             Map<String, String> serviceLabels = sdSvc.getServiceLabels(service);
             if (serviceLabels.containsKey(ServiceDiscoveryConstants.LABEL_SERVICE_GLOBAL) &&
                     allocatorSvc.hostSatisfiesHostAffinity(hostId, serviceLabels)) {
-                activate(service);
+                DeferredUtils.defer(new DeferredServiceActivate(service));
```

Comment:
Instead of this pattern, I think you want to io.cattle.platform.servicediscovery.deployment.DeploymentManager#reconcileServicesFor() which will schedule a DeploymentManager.activate() in the background in a reliable fashion.  for this to work though you need to change io.cattle.platform.servicediscovery.service.ServiceDiscoveryService#getServicesFor() also to handle objects of instance type `Service`